### PR TITLE
AAD TokenCredential support on EventHub Stream Provider

### PR DESCRIPTION
**Pull request by galvesribeiro**
Backport #6667 to 3.x

### CODE REVIEW

File: src/Azure/Orleans.Streaming.EventHubs/Providers/EventDataGeneratorStreamProvider/EventHubPartitionDataGenerator.cs [hunk 1]
```diff
@@ -54,22 +54,25 @@ public bool TryReadEvents(int maxCount, out IEnumerable<EventData> events)
...
+                    this.StreamId.Namespace,
+                    this.GenerateEvent(this.SequenceNumberCounter.Value),
+                    RequestContextExtensions.Export(this.serializationManager));
+
+                var wrapper = new WrappedEventData(
+                    eventData.Body,
+                    eventData.Properties,
+                    eventData.SystemProperties,
+                    partitionKey: this.StreamId.Guid.ToString(),
+                    offset: DateTime.UtcNow.Ticks,
```
Comment (benjaminpetit) [inline src/Azure/Orleans.Streaming.EventHubs/Providers/EventDataGeneratorStreamProvider/EventHubPartitionDataGenerator.cs:70]: My main concern about this change is the offset. Is it a breaking change?

**Comment by benjaminpetit** [comment]
Before merging this PR, I will do some extensive tests so we can be condifent that we are not introducing a breaking change in the streaming system

File: src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/IEventHubReceiver.cs [hunk 2]
```diff
@@ -33,21 +34,61 @@ public interface IEventHubReceiver
...
+                 : new PartitionReceiver(options.ConsumerGroup, partitionSettings.Partition, GetEventPosition(), options.ConnectionString, options.Path, receiverOptions);
+
+             EventPosition GetEventPosition()
+             {
+                 EventPosition eventPosition;
+
+                 // If we have a starting offset, read from offset
+                 if (offset != EventHubConstants.StartOfStream)
+                 {
+                     if (!long.TryParse(offset, out var longOffset))
```
Comment (benjaminpetit) [inline src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/IEventHubReceiver.cs:59]: I think we already check that in `EventHubCheckpointer.Load`, but it is still good do to it here. But I think we should log this as an error, and then treat as there was no checkpoint in storage (and respect the value of `ReceiverOptions.StartFromNow`)

File: src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/EventHubStreamOptions.cs [hunk 3]
```diff
@@ -23,6 +23,15 @@ public class EventHubOptions
         /// Hub path.
         /// </summary>
         public string Path { get; set; }
+        /// <summary>
+        /// The token credential.
+        /// </summary>
+        public TokenCredential TokenCredential { get; set; }
```
Comment (benjaminpetit) [inline src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/EventHubStreamOptions.cs:29]: If this is a secret, we should add a `[Redact]' attribute so it is not logged at silo startup

File: src/Azure/Orleans.Streaming.EventHubs/Providers/EventDataGeneratorStreamProvider/EventHubPartitionDataGenerator.cs (hunk 1 above)
Comment (galvesribeiro) [reply src/Azure/Orleans.Streaming.EventHubs/Providers/EventDataGeneratorStreamProvider/EventHubPartitionDataGenerator.cs:70]: Nop, it is behaving the same way as the previous one. In fact I haven't change the behavior at all. Just got the PR that is merged in master and adapted it to fit the old SDK on 3.x.

File: src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/IEventHubReceiver.cs (hunk 2 above)
Comment (galvesribeiro) [reply src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/IEventHubReceiver.cs:59]: hummm make sense. I Just tried to not change the way it was implemented on master. I would rather prefer to add a PR later to enhance this and keep this one as close as possible to the PR we already have merged.

File: src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/EventHubStreamOptions.cs (hunk 3 above)
Comment (galvesribeiro) [reply src/Azure/Orleans.Streaming.EventHubs/Providers/Streams/EventHub/EventHubStreamOptions.cs:29]: Should not be a secret. This is the runtime account/principal used. It has a `GetTokenAsync()` method that then gets the access_token for the resources.

**Pull request merged**